cmd/kanikoExecute.go

```diff
@@ -0,0 +1,122 @@
+	c := command.Command{
+		ErrorCategoryMapping: map[string][]string{
+			log.ErrorConfiguration.String(): []string{
+				"unsupported status code 401",
+			},
+		},
+	}
+
+	// reroute command output to logging framework
+	c.Stdout(log.Writer())
+	c.Stderr(log.Writer())
+
+	client := &piperhttp.Client{}
+
+	fileUtils := &piperutils.Files{}
+
+	err := runKanikoExecute(&config, telemetryData, &c, client, fileUtils)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("Kaniko execution failed")
+	}
+}
+
+func runKanikoExecute(config *kanikoExecuteOptions, telemetryData *telemetry.CustomData, execRunner command.ExecRunner, httpClient piperhttp.Sender, fileUtils piperutils.FileUtils) error {
+	// backward compatibility for parameter ContainerBuildOptions
+	if len(config.ContainerBuildOptions) > 0 {
+		config.BuildOptions = strings.Split(config.ContainerBuildOptions, " ")
+		log.Entry().Warning("Parameter containerBuildOptions is deprecated, please use buildOptions instead.")
+		telemetryData.Custom1Label = "ContainerBuildOptions"
+		telemetryData.Custom1 = config.ContainerBuildOptions
+	}
+
+	// prepare kaniko container for running with proper Docker config.json and custom certificates
+	// custom certificates will be downloaded and appended to ca-certificates.crt file used in container
+	prepCommand := strings.Split(config.ContainerPreparationCommand, " ")
+	if err := execRunner.RunExecutable(prepCommand[0], prepCommand[1:]...); err != nil {
+		return errors.Wrap(err, "failed to initialize Kaniko container")
+	}
+
+	err := certificateUpdate(config.CustomTLSCertificateLinks, httpClient, fileUtils)
+	if err != nil {
+		return errors.Wrap(err, "failed to update certificates")
+	}
+
+	if !piperutils.ContainsString(config.BuildOptions, "--destination") {
+		dest := []string{"--no-push"}
+		if len(config.ContainerImage) > 0 {
+			dest = []string{"--destination", config.ContainerImage}
+		}
+		config.BuildOptions = append(config.BuildOptions, dest...)
+	}
+
+	dockerConfig := []byte(`{"auths":{}}`)
+	if len(config.DockerConfigJSON) > 0 {
+		dockerConfig, err = fileUtils.FileRead(config.DockerConfigJSON)
+		if err != nil {
+			return errors.Wrapf(err, "failed to read file '%v'", config.DockerConfigJSON)
+		}
+	}
+
+	if err := fileUtils.FileWrite("/kaniko/.docker/config.json", dockerConfig, 0644); err != nil {
+		return errors.Wrap(err, "failed to write file '/kaniko/.docker/config.json'")
+	}
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		return errors.Wrap(err, "failed to get current working directory")
+	}
+	kanikoOpts := []string{"--dockerfile", config.DockerfilePath, "--context", cwd}
+	kanikoOpts = append(kanikoOpts, config.BuildOptions...)
+
+	err = execRunner.RunExecutable("/kaniko/executor", kanikoOpts...)
+	if err != nil {
+		return errors.Wrap(err, "execution of '/kaniko/executor' failed")
+	}
+	return nil
+}
+
+func certificateUpdate(certLinks []string, httpClient piperhttp.Sender, fileUtils piperutils.FileUtils) error {
```

Comment:
As this may be used in multiple steps, certificate handling could be moved to dedicated package.